modules/concurrency/include/hephaestus/concurrency/repeat_until.h

```diff
@@ -5,117 +5,131 @@
+  using SenderT = std::invoke_result_t<SenderFactory>;
 
-namespace internal {
-template <typename SenderFactoryT, typename ReceiverT>
-struct RepeatUntilStateT {
-  using SenderT = std::invoke_result_t<SenderFactoryT>;
-  struct InnerReceiverT {
-    RepeatUntilStateT* self;
-    using receiver_concept = stdexec::receiver_t;
-
-    // NOLINTNEXTLINE(readability-identifier-naming)
-    void set_value(bool done) const noexcept {
+  template <typename Receiver>
+  struct Operation {
+    struct InnerReceiver {
+      using receiver_concept = stdexec::receiver_t;
+      Operation* self;
+
+      // NOLINTBEGIN(readability-identifier-naming) - wrapping stdexec interface
+      void set_value(bool done) const noexcept {
+        self->next(done);
+      }
+
+      void set_stopped() const noexcept {
+        self->setStopped();
+      }
+
+      template <typename Error>
+      void set_error(Error error) const noexcept {
+        self->setError(std::move(error));
+      }
+
+      auto get_env() const noexcept {
+        return stdexec::get_env(self->receiver);
+      }
+      // NOLINTEND(readability-identifier-naming)
+    };
+
+    SenderFactory sender_factory;
+    Receiver receiver;
+    stdexec::__optional<stdexec::connect_result_t<SenderT, InnerReceiver>> state;
+
+    void start() noexcept {
+      next(false);
+    }
+
+    void next(bool done) {
       if (done) {
-        stdexec::set_value(std::move(self->receiver));
+        stdexec::set_value(std::move(receiver));
+        return;
+      }
+      try {
+        state.emplace(stdexec::__emplace_from{
+            [this]() { return stdexec::connect(sender_factory(), InnerReceiver{ this }); } });
+
+      } catch (...) {
+        stdexec::set_error(std::move(receiver), std::current_exception());
         return;
       }
-      self->start();
+      stdexec::start(*state);
     }
 
-    // NOLINTNEXTLINE(readability-identifier-naming)
-    void set_stopped() const noexcept {
-      stdexec::set_stopped(std::move(self->receiver));
+    void setStopped() {
+      stdexec::set_stopped(std::move(receiver));
     }
 
     template <typename Error>
-    // NOLINTNEXTLINE(readability-identifier-naming)
-    void set_error(Error error) const noexcept {
+    void setError(Error error) noexcept {
       if constexpr (std::is_same_v<std::decay_t<Error>, std::exception_ptr>) {
-        stdexec::set_error(std::move(self->receiver), std::move(error));
+        stdexec::set_error(std::move(receiver), std::move(error));
+      }
+      if constexpr (std::is_same_v<std::decay_t<Error>, std::error_code>) {
+        stdexec::set_error(std::move(receiver), std::make_exception_ptr(std::system_error(std::move(error))));
       } else {
-        try {
-          throw error;
-        } catch (...) {
-          stdexec::set_error(std::move(self->receiver), std::current_exception());
-        }
+        stdexec::set_error(std::move(receiver), std::make_exception_ptr(std::move(error)));
       }
```

Comment:
BTW, if we have this pattern in more places in the source code, might be worthwhile refactoring to a `normalizeError` function or something similar.